### DISCUSSION TITLE
Note

### DISCUSSION THREAD

**Discussion by pixincreate** [pixincreate/GMSCompat#5] [General]
First of all, I'm really new to this! I might many mistakes and I hope I learn a lot through this. This mGMSCompat is a high-priority project for me to do.  

- mGMSCompat is a Magisk module that as usual, requires you to have root access in order for it to work.  
- My plans are pretty straight forward as described below:
  - Installing this module should install @GrapheneOS [Apps](https://github.com/GrapheneOS/Apps) and GMSCompat as system apps which takes care of GSF, GMS, GPlay and GMSCompat (config) in 1 go. 
  - And yes, all the 4 apps are required for some apps to work. 
- I don't think it would take much efforts to add their app store that can be the appStore for these apps. 
- The actual challenge lies in the below given parts:
  - Modifying the system-partition with Magisk in systemless mode that adds support for GMS sandboxing. 
    - This AFAIK requires me to make modifications to the source code. Since the code is already compiled and running at the time of installation, I should resort to other methods that modifies makes the changes such as replacing the files if that's supported. 
  - After doing some homework, I learnt that we can pass Play Integrity API (PIA) without the necessity of locking the bootloader and I've no interest in integrating @kdrag0n's [safetynet-fix](https://github.com/kdrag0n/safetynet-fix) module within mGMSCompat as it will lead to unnecessary complications. 

> ## Few more thoughts on GMSCompat after using GrapheneOS 
>
> - mGMSCompat requires root access to work while GOS requires the bootloader to be locked to pass the Play Integrity API.
> - I'm not sure whether its the ROM side thing or something different. What I mean by this is that if locking bootloader is necessary after installing GMSCompat, then it defeats the purpose of this project. 
> - Looking forward to learn more things regarding this.

This was in my readme.md until [this](https://github.com/pixincreate/mGMSCompat/commit/4d5e8f9b8d9095e3c5293a02cc93613515b2258a) commit and now I believe I have found a convincing answer to that. @GrapheneOS has it's own signing key, requirements and all those stuffs that makes the ROM fully functional **only** if the user has **locked** the bootloader viz., for PIA to pass, verified boot is mandatory on GOS. This not only passes the PIA but also many other security features that greatly enhances the privacy of an individual.  
This would also make back-porting support for older Android devices a lot more harder and even increase the size of the module that would increase the module load time which I've no interest in. May be mandating the internet access at the time of installation would solve the issue while giving user an option to backup the apps similar to what [Pixelify](https://github.com/Kingsman44/Pixelify) does.

Also, @SkewedZeppelin has [rejected](https://github.com/Divested-Mobile/DivestOS-Build/discussions/22) the request for adding GMSCompat support for DivestOS project. So mGMSCompat would be a better fit that would help users who want GrapheneOS hardening features, root and GMSCompat available on a single place. 

> Had a small talk with @KeironQuinn the other day to learn that I need either Zygisk or LSPosed to achieve this. Working with LSPosed is easy compared to directly putting hands into Zygisk. With that said, I'm currently planning to go with LSPosed and later, once I'm good with this, will slowly migrate to Zygisk.

## For reference

- [VoltageOS/packages_apps_GmsCompat](https://github.com/VoltageOS/packages_apps_GmsCompat)
- [VoltageOS/external_GmsCompatConfig](https://github.com/VoltageOS/external_GmsCompatConfig)
- [VoltageOS/frameworks_base](https://github.com/VoltageOS/frameworks_base)
- [SandboxedGooglePlayCompatibilityLayerCommits](https://gist.github.com/thestinger/ee536cbd1ca674b94dde05831192c348)
- https://github.com/GrapheneOS/platform_external_Apps
- https://github.com/GrapheneOS/platform_external_GmsCompatConfig
- https://github.com/GrapheneOS/platform_packages_apps_GmsCompat

**Comment by eugenio412** [comment]
Hello @pixincreate ! great idea! this would be a super nice feature that I would like to have! did you manage to advance on your project? was it tested somewhere?

**Reply by pixincreate** [reply]
Hi @eugenio412, I'm running late actually and couldn't focus on this project for some time. Will be back on it this weekend.

**Comment by lepras** [comment]
I think the best way will be to use KernelSU (Hide root from apps), then Zygisk, then LSposed.

We cracked the play integrity here:

https://forum.xda-developers.com/t/info-play-integrity-api-replacement-for-safetynet.4479337/

I have documented banking apps on AOSP [here](https://github.com/lepras/bankingAOSP/)

**Reply by pixincreate** [reply]
Hey @lepras, thanks for letting me know about this. The purpose of this project is allow users to install GApps without baking that into ROM by modifying some parts of Vanilla ROM by means of Magisk.   
Unfortunately, I'm unable focus on this currently since I cannot leverage my daily driver due to lack of time.

**Reply by lepras** [reply]
I am also interested in the same, I have checked your links, its only code. Do you have any links that discuss the design of the GMScompat?

Probably they are using a custom android profile?

I will dig in the code if I get the time.

**Reply by pixincreate** [reply]
I haven't looked at the discussion that involved the design, but rather had a communication with the developers and documented some of the things that I remembered
